theseus/optimizer/nonlinear/trust_region.py

```diff
@@ -0,0 +1,167 @@
+#       - expand_threshold (default 0.75)
+#       - shrink_ratio (default 0.25)
+#       - expand_ratio (default 2.0)
+#
+# Iff the ratio between the actual error reduction after the step and the
+# predicted one is greater than `accept_threshold`, the step is applied.
+# Iff it's also greater than `expand_threshold`, the trust region is also
+# increased by `expand_ratio`.
+# Iff the ratio is lower than `shrink_threshold`,
+# then the trust region is reduced by `shrink_ratio`.
+# The trust region is initialized to `trust_region_init`.
+class TrustRegion(NonlinearLeastSquares, abc.ABC):
+    _MIN_TRUST_REGION = 1.0e-5
+    _MAX_TRUST_REGION = 1.0e5
+
+    def __init__(
+        self,
+        objective: Objective,
+        linear_solver_cls: Optional[Type[LinearSolver]] = None,
+        vectorize: bool = False,
+        linearization_cls: Optional[Type[Linearization]] = None,
+        linearization_kwargs: Optional[Dict[str, Any]] = None,
+        linear_solver_kwargs: Optional[Dict[str, Any]] = None,
+        abs_err_tolerance: float = 1e-10,
+        rel_err_tolerance: float = 1e-8,
+        max_iterations: int = 20,
+        step_size: float = 1.0,
+        **kwargs,
+    ):
+        super().__init__(
+            objective,
+            linear_solver_cls=linear_solver_cls,
+            vectorize=vectorize,
+            linearization_cls=linearization_cls,
+            linearization_kwargs=linearization_kwargs,
+            linear_solver_kwargs=linear_solver_kwargs,
+            abs_err_tolerance=abs_err_tolerance,
+            rel_err_tolerance=rel_err_tolerance,
+            max_iterations=max_iterations,
+            step_size=step_size,
+            **kwargs,
+        )
+        self._trust_region: torch.Tensor = None
+        self._at_trust_boundary_idx: torch.Tensor = None
+
+    def reset(
+        self,
+        trust_region_init: float = 0.5,
+        **kwargs,
+    ) -> None:
+        super().reset(**kwargs)
+        self._trust_region = trust_region_init * torch.ones(
+            self.objective.batch_size,
+            1,
+            device=self.objective.device,
+            dtype=self.objective.dtype,
+        )
+        self._at_trust_boundary_idx = None
+
+    # Return the computed delta and, optionally, the indices that with the
+    # steps that are exactly at the boundary of the trust region
+    @abc.abstractmethod
+    def _compute_delta_impl(self) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
+        pass
+
+    def compute_delta(self, **kwargs) -> torch.Tensor:
+        # Storing the indices at the trust boundary so that we can update the
+        # trust region inside self.complete_step()
+        delta, self._trusted_step_idx = self._compute_delta_impl()
+        if self._trusted_step_idx is None:
+            self._trusted_step_idx = (
+                TrustRegion._squared_norm(delta) <= self._trust_region**2
+            )
+
+        # Storing the indices of delta exactly at the trust boundary.
```

Comment:
Not sure if we need to store the indices of delta exactly at the trust boundary since the expansion and shrinking of the trust region usually only depends on whether the current linearized model gives a good approximation.

Reply:
See comment below.